engine/src/main/java/org/pentaho/di/trans/steps/mergejoin/MergeJoin.java

```diff
@@ -0,0 +1,466 @@
+           */
+          data.two = getRowFrom( data.twoRowSet );
+        }
+        break;
+      case -1:
+        // if (log.isDebug()) logDebug("Second stream has missing key");
+        /*
+         * Second stream is greater than the first stream. This means: a) This key is missing in the second stream b)
+         * First stream may have finished So, if full/left outer join is set and 1st stream is not null, we push a
+         * record to output with only the values for the first row populated. Next, if 1st stream is not finished, we
+         * get a row from it; otherwise signal that we are done
+         */
+        if ( data.two_optional == true ) {
+          if ( data.one != null ) {
+            Object[] outputRowData = RowDataUtil.createResizedCopy( data.one, data.outputRowMeta.size() );
+            outputRowData = RowDataUtil.addRowData( outputRowData, data.oneMeta.size(), data.two_dummy );
+            putRow( data.outputRowMeta, outputRowData );
+            data.one = getRowFrom( data.oneRowSet );
+          } else if ( data.one_optional == false ) {
+            /*
+             * We are doing a left outer join and there are no more rows in the first stream; so we are done
+             */
+            // Before we stop processing, we have to make sure that all rows from both input streams are depleted!
+            // If we don't do this, the transformation can stall.
+            //
+            while ( data.one != null && !isStopped() ) {
+              data.one = getRowFrom( data.oneRowSet );
+            }
+            while ( data.two != null && !isStopped() ) {
+              data.two = getRowFrom( data.twoRowSet );
+            }
+
+            setOutputDone();
+            return false;
+          } else {
+            /*
+             * We are doing a full outer join so print the 2nd stream and get the next row from the 2nd stream
+             */
+            Object[] outputRowData = RowDataUtil.createResizedCopy( data.one_dummy, data.outputRowMeta.size() );
+            outputRowData = RowDataUtil.addRowData( outputRowData, data.oneMeta.size(), data.two );
+            putRow( data.outputRowMeta, outputRowData );
+            data.two = getRowFrom( data.twoRowSet );
+          }
+        } else if ( data.one == null && data.one_optional == true ) {
+          /*
+           * We have reached the end of stream 1 and there are records present in the second stream. Also, join is right
+           * or full outer. So, create a row with just the values in the 2nd stream and push it forward
+           */
+          Object[] outputRowData = RowDataUtil.createResizedCopy( data.one_dummy, data.outputRowMeta.size() );
+          outputRowData = RowDataUtil.addRowData( outputRowData, data.oneMeta.size(), data.two );
+          putRow( data.outputRowMeta, outputRowData );
+          data.two = getRowFrom( data.twoRowSet );
+        } else if ( data.one != null ) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

It seems unusual to set the value of a variable only when it is already set. You might want to switch the null check of != on line 376 to == check before setting the field value on line 381 or verify if the check is needed at all.